### PR TITLE
[mlir][tensor] Fold unpadding collapse_shape into extract_slice

### PR DESCRIPTION
**Pull request by adam-smnk**

### CODE REVIEW

File: mlir/test/Dialect/Tensor/fold-reassociative-reshapes.mlir [hunk 1]
```diff
@@ -22,6 +22,75 @@ func.func @expand_shape_of_rank_reducing_extract(
 
 // -----
 
+// CHECK-LABEL: func @unpadding_collapse_of_extract_slice(
+//  CHECK-SAME:     %[[t:.*]]: tensor<?x?x?x?xf32>
+//  CHECK-SAME:     %[[x:[a-zA-Z0-9_]+]]: index
+//  CHECK-SAME:     %[[y:[a-zA-Z0-9_]+]]: index
+//       CHECK:   %[[extract:.*]] = tensor.extract_slice %[[t]][%[[x]], %[[y]], 0, 0] [1, %{{.*}}, 1, %{{.*}}] [1, 1, 1, 1] : tensor<?x?x?x?xf32> to tensor<?x?xf32>
```
Comment (qedawkins) [inline mlir/test/Dialect/Tensor/fold-reassociative-reshapes.mlir:29]: nit: Can you reduce the line length by splitting this into two lines?
```suggestion
//       CHECK:   %[[extract:.*]] = tensor.extract_slice %[[t]][%[[x]], %[[y]], 0, 0] [1, %{{.*}}, 1, %{{.*}}] [1, 1, 1, 1]
//  CHECK-SAME:     : tensor<?x?x?x?xf32> to tensor<?x?xf32>
```
Same for the two tests below.

File: mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp [hunk 2]
```diff
@@ -48,6 +48,39 @@ struct FoldExpandOfRankReducingExtract
...
+  using OpRewritePattern<tensor::CollapseShapeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::CollapseShapeOp collapseShapeOp,
+                                PatternRewriter &rewriter) const override {
+    auto extractSliceOp =
+        collapseShapeOp.getSrc().getDefiningOp<tensor::ExtractSliceOp>();
+    // Collapse cannot be folded away with multiple users of the extract slice
+    // and it is not necessarily beneficial to only convert the collapse into
+    // another extract slice.
+    if (!extractSliceOp || !extractSliceOp.getResult().hasOneUse())
```
Comment (qedawkins) [inline mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:64]: nit:
```suggestion
    if (!extractSliceOp || !extractSliceOp.hasOneUse())
```

**Review by qedawkins — APPROVED** [review]
LGTM, thanks!

**Review by chelini — APPROVED** [review]
Looks good to me as well.

**Pull request merged**